windows/src/engine/keyman/UfrmKeyman7Main.pas

```diff
@@ -983,60 +983,136 @@ procedure TfrmKeyman7Main.LanguageSwitchFormHidden(Sender: TObject);
+      // This may never happen as it may have been a VCL bug, but I'm including
+      // it just for safety given it can return nil.
+      Exit(alBottom);
+
+    if r.Left = m.Left then
+    begin
+      if r.Top = m.Top then
+      begin
+        if r.Right = m.BoundsRect.Right then
+          Exit(alTop);
+        Exit(alLeft);
+      end;
+      Exit(alBottom)
+    end;
+    Exit(alRight);
+  end;
+
+  function GetMenuAnchorPoint(systraylocation: TAlign): TPoint;
+  type
+    TGetDpiForWindow = function(hwnd: HWND): UINT; stdcall;
+  var
+    pt: TPoint;
+    r: TRect;
+    ident: NOTIFYICONIDENTIFIER;
+    dpi: Cardinal;
+    m: TMonitor;
+    FGetDpiForWindow: TGetDpiForWindow;
+  begin
+    // We'll try and get the actual location of our icon
+    ident.cbSize := SizeOf(ident);
+    ident.hWnd := FRunningProduct.FTrayIcon.NotificationWindow;
+    ident.uID := 1;
+    ident.guidItem := GUID_NULL;
+    if TriggeredByKeyboard or (Shell_NotifyIconGetRect(ident, r) <> S_OK) then
+    begin
+      // User has requested the menu with a hotkey, so show it near the
+      // SystemNotificationArea, but not actually where our icon is, because
+      // the icon may be hidden (or if we fail to get the location of our icon,
+      // for any reason)
+      if not GetWindowRect(hwndSysTray, r) then
+        r := TRect.Empty;
+    end
+    else
+    begin
+      // Shell_NotifyIconGetRect does not honour the DPI awareness of the caller
+      // and as we do not yet support per-monitor DPI awareness, we have to
+      // scale the rectangle by its actual DPI to our fixed 96 DPI value to
+      // get the location of the rect in our 96 DPI world. As a monitor may have
+      // a non-zero origin, we have to deal with that too!
+
+      // Don't use Delphi's GetDpiForWindow declaration as we have to verify
+      // Windows version first, which cascades into potential updates to the
+      // manifest, which we don't want to do just now.
+      FGetDpiForWindow := GetProcAddress(GetModuleHandle(user32), 'GetDpiForWindow');
+      if Assigned(FGetDpiForWindow) then
+      begin
+        dpi := FGetDpiForWindow(hwndTray);
+        m := Screen.MonitorFromWindow(hwndTray);
+        if Assigned(m) then
+        begin
+          r.Right := MulDiv(r.Right-m.Left, 96, dpi) + m.Left;
+          r.Bottom := MulDiv(r.Bottom-m.Top, 96, dpi) + m.Top;
+          r.Left := MulDiv(r.Left-m.Left, 96, dpi) + m.Left;
+          r.Top := MulDiv(r.Top-m.Top, 96, dpi) + m.Top;
+        end
+        else if not GetWindowRect(hwndSysTray, r) then
+          r := TRect.Empty;
+      end;
+    end;
+
+    // Align the menu according to the location of the taskbar
+    case systraylocation of
+      alTop:    begin pt.X := (r.Right + r.Left) div 2; pt.Y := r.Bottom; end;
+      alBottom: begin pt.X := (r.Right + r.Left) div 2; pt.Y := r.Top; end;
+      alLeft:   begin pt.X := r.Right; pt.Y := r.Top; end;
+      alRight:  begin pt.X := r.Left; pt.Y := r.Top; end;
+    end;
```

Comment:
This seems pretty close to windows/src/engine/keyman/UfrmKeymanMenu.pas, line 478+.  Any reason for the WETness here?

Reply:
This here calculates the anchor point of the menu; the code in UfrmKeymanMenu calculates the actual position of the menu with reference to the anchor point, once we know its width and height.

We could move these changes into the UfrmKeymanMenu unit but the calculations are different in each case so it's not going to change much.